modules/mod-deep-learning/DeepModel.cpp

```diff
@@ -0,0 +1,199 @@
+
+      // set mModel
+      mModel = std::make_unique<torch::jit::script::Module>
+                  (torch::jit::load(modelPath, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error while loading model"), e.what());
+   }
+}
+
+void DeepModel::Load(std::istream &bytes)
+{
+   try
+   {
+      // load the resampler module
+      LoadResampler();
+
+      // load the model to CPU, as well as the metadata
+      mModel = std::make_unique<torch::jit::script::Module>
+                     (torch::jit::load(bytes, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error while loading model"), e.what());
+   }
+}
+
+bool DeepModel::IsLoaded() const
+{
+   return mLoaded; 
+}
+
+void DeepModel::Save(const std::string &modelPath) const
+{
+   if (!mLoaded)
+      throw ModelException(XO("attempted save when no module was loaded."), "");
+
+   mModel->save(modelPath);
+}
+
+void DeepModel::SetCard(ModelCardHolder card)
+{
+   // set the card
+   mCard = card;
+
+   // set sample rate
+   mSampleRate = mCard->sample_rate();
+}
+
+ModelCardHolder DeepModel::GetCard() const
+{
+   return ModelCardHolder(mCard);
+}
+
+void DeepModel::Cleanup()
+{
+   // cleanup
+   mModel.reset();
+   mLoaded = false;
+}
+
+torch::Tensor DeepModel::Resample(const torch::Tensor &waveform, int sampleRateIn, 
+                                  int sampleRateOut) const
+{
+   if (!mLoaded) 
+      throw ModelException(XO("Attempted resample while is not loaded."
+                                       " Please call Load() first."), ""); 
+
+   // set up inputs
+   // torchaudio likes that sample rates are cast to float, for some reason.
+   std::vector<torch::jit::IValue> inputs = {waveform, 
+                                             static_cast<float>(sampleRateIn), 
+                                             static_cast<float>(sampleRateOut)};
+
+   try
+   {
+      return mResampler->forward(inputs).toTensor();
+   }
+   catch (const std::exception &e)
+   {
+      throw ModelException(XO("A libtorch error occured while resampling."), e.what());
+   }
+}
+
+// forward pass through the model!
+torch::jit::IValue DeepModel::Forward(const torch::Tensor &waveform) const
+{
+   // NoGradGuard prevets the model from storing gradients, which should
+   // make computation faster and memory usage lower. 
+   torch::NoGradGuard NoGrad;
+
+   if (!mLoaded) 
+      throw ModelException(XO("Attempted forward pass while model is not loaded."
+                                       " Please call Load() first."), ""); 
+
+   // set up for jit model
+   std::vector<torch::jit::IValue> inputs = {waveform};
+
+   // forward pass!
+   try
+   {
+      return  mModel->forward(inputs);
```

Comment:
Is this truly strongly exception-safe, meaning that if `forward` throws, `*mModel` is in an unchanged state?  To answer that would require knowing what promises Torch makes.